### DISCUSSION TITLE
Position smaller cropped parts to (larger) original image - composite crop join geometry coordinates

### DISCUSSION THREAD

**Discussion by fred888651** [ImageMagick/ImageMagick#2603] [Help] [answered]
ImageMagick-7.0.9-Q16
Windows 10 

**QUESTION: How do I place one (or more) small cropped parts onto larger (original) image?
Now it seems to place itself in top left corner (0:0 xy coordinate) - it should place itself in its "original" position.**

I have tried -gravity center etc - but I think I need someway to send exact coordinates to place the small cropped part/parts - in its "original" x:y position? How would one do that? -geometry x:y ?

(Cropped parts output being done using mogrify crop command)

![image](https://user-images.githubusercontent.com/70697839/93747053-6b5cec80-fbf6-11ea-8248-d0fec61300ce.png)

**Comment by snibgo** [comment]
When you "-crop" an image, IM records page.x and page.y, the offsets of the crop in the original image. "+repage" removes this metadata, so don't do that. If you write this crop to a file, some formats such as PNG will record the offsets but some such as JPG won't.

If you use a "-layers" operation, those offsets will be used.

For example:
```
magick in.png -crop 2x3@ +write info: -layers flatten out.png
```
This divides the input into 6 pieces, then immediately reassembles them, so out.png is the same as in.png. This is pointless, but may help you understand what is happening.

**Reply by fred888651** [reply]
Thanks again Snigbo!
So the -layers option uses the original x:y position of the cropped part 
That is very interesting
Removed the +repage and changed format to .PNG 

However this created a new problem in my script.
Have to do more tinkering :)
(works perfect with .JPG oddly enough. (split into 36 parts)
Well well, that is the nature of it, if you change a half a comma in the code - surprises will come :) )


I am trying to write a script that:

1. first crops original .PNG in 6x6=36 files ( standard portrait/tall 2:3 image) 

2. randomly selects an arbitrary amount of cropped smaller parts

3. ..."glue" them back on to the blurred original photo  

4.  loop this a number of times to generate a batch of randomly different composite files 

5. read all these composite files and output animation

**Comment by GeeMack** [comment, marked as answer]
Using IMv7 on Windows 10, this command will do pretty much what you've described... 

    magick lena.jpg -alpha set -background none -write mpr:img -blur 0x12 null: -channel A ^
       ( mpr:img -crop 6x6@ -evaluate multiply %[fx:floor(rand()*6)?0:1] -flatten ) ^
       ( mpr:img -crop 6x6@ -evaluate multiply %[fx:floor(rand()*6)?0:1] -flatten ) ^
       ( mpr:img -crop 6x6@ -evaluate multiply %[fx:floor(rand()*6)?0:1] -flatten ) ^
       ( mpr:img -crop 6x6@ -evaluate multiply %[fx:floor(rand()*6)?0:1] -flatten ) ^
       ( mpr:img -crop 6x6@ -evaluate multiply %[fx:floor(rand()*6)?0:1] -flatten ) ^
       ( mpr:img -crop 6x6@ -evaluate multiply %[fx:floor(rand()*6)?0:1] -flatten ) ^
       ( mpr:img -crop 6x6@ -evaluate multiply %[fx:floor(rand()*6)?0:1] -flatten ) ^
       ( mpr:img -crop 6x6@ -evaluate multiply %[fx:floor(rand()*6)?0:1] -flatten ) ^
       ( mpr:img -crop 6x6@ -evaluate multiply %[fx:floor(rand()*6)?0:1] -flatten ) ^
       ( mpr:img -crop 6x6@ -evaluate multiply %[fx:floor(rand()*6)?0:1] -flatten ) ^
       +channel -layers composite -set delay 20 -loop 0 output.gif

![output](https://user-images.githubusercontent.com/65475279/93811641-1204b580-fc16-11ea-8192-67790086fd03.gif)

First it reads in the original image, puts a copy in memory with "-write mpr:img", and blurs the original.

The "-channel A" means everything that follows until the "+channel" will only affect the alpha channel.

Then inside the first parentheses it pulls in a copy of the input "mpr:img" and crops it into 6 rows by 6 columns, or 36 equal pieces.

Next the "-evaluate" operation uses an FX expression to change the alpha value to transparent for a random choice of about one of six of the pieces. It composites the pieces in their original locations onto a transparent background. The transparent pieces just look like holes or empty spots in the image.

Then in more sets of parentheses we do that same process of cropping, making random transparent pieces, and re-assembling nine more times so we have 10 images with randomly deleted holes.

After creating all the images with transparent holes we composite them one at a time with "-layers composite" over the blurred input image to make the 10 frames for the output. Set the delay for the speed of the GIF, and write the output animation.

If you're curious about the workings of some of these operations, you can put "-write testing%02d.png" pretty much anywhere along the command, run the command, and there will be one or more "testingNN.png" file(s) you can look at to see what's happening.

Add more of the "(...)" lines to make more frames.

To make more or fewer "holes" in the frames, change the 6 in that FX expression "%[fx:floor(rand()*6)?0:1]" to your desired ratio of "holes".

If you're running this from a BAT script you'll need to double all the percent signs "%%".

That requires sufficient memory to read the original image and create as many frames as you need for your animation, so this might not be the best solution if you have large images or lots of frames.

**Reply by fred888651** [reply]
Thanks! This is great stuff!

Been tinkering around with it a bit and it does exactly what I was looking for!

Im still learning all this new IM stuff - not easy :) 

-alpha is the channel for transparency

-write writes image sequence to MPR - where MPR is some kind of in RAM memory storage - for faster pipeline operations

-evaluate multiply - I believe this is what Photoshop calls blending mode (multiply) ?

%[ fx:floor (rand()*6) ?0:1 ] random number 0-6, rounded down
I dont understand the "?0:1" part - tried to search it but just found ? meaning a ternary operator - but if so it should have 3 values? 

-flatten means add prev+current image on to each other 

-composite is putting all previous frames together in specified formar .GIF/.MP4/.AVI etc - delay is delay and -loop 0 means repeat forever

I guess you also could write all these 
_( mpr:img -crop 6x6@ -evaluate multiply %[fx:floor(rand()*6)?0:1] -flatten )_
in a for-loop ?

**Reply by GeeMack** [reply]
The "-alpha set" makes sure the image has an alpha channel.

You can use "-write" to put one or more images in a memory register that can be accessed until the end of the command. You can also use "-write" to write intermediate files to disk like this...

    magick wizard: -resize 50% -write 50percent.png -resize 50% -write 25percent.png -resize 50% verysmall.png

The "-evaluate" operation can do many things, but in this case it uses "multiply" to multiply the alpha channel by 0 or 1 depending on the value of the FX expression. Multiplying the alpha channel by 0 makes it transparent. Multiplying it by 1 leaves it unchanged.

That particular FX expression works like if-then-else. The "floor(rand()*6)" part selects a random whole number between 0 and 5. The question mark "?" checks that first statement for true or false. In this case a 0 would be false, and any other number would be true. After the question mark is the return value *if* the statement is true, 1, 2, 3, 4, or 5, which will be roughly 5/6 of the time here. Then after the colon ":" is the value to return if the statement is false.

So in this command it multiplies the value of the alpha channel by 0 to make it transparent about 84% of the time.

Both "-flatten" and "-composite" are operators to assemble one or more layers on top of each other in various ways. In this case we use "-layers composite", which is a special way to composite each of several images, one at a time, onto a single base image and output multiple results. Note the "null:", a special IM file format required here to indicate the split between the sets of source and destination layers to be composited.

Something like this would work within a "for" loop to create one image at a time...

    magick lena.jpg -alpha set -background none -write mpr:img -blur 0x12 -channel A ^
       ( mpr:img -crop 6x6@ -evaluate multiply %[fx:floor(rand()*6)?0:1] -flatten ) ^
       +channel -composite result.png

Keep in mind IM will overwrite existing files happily and without warning, so to loop a command like this requires some way to handle incrementing the output file names.
